drivers/hmis_external_apis/app/jobs/hmis_external_apis/create_referral_job.rb

```diff
@@ -7,78 +7,70 @@
 module HmisExternalApis
   class CreateReferralJob < ApplicationJob
     include HmisExternalApis::ReferralJobMixin
-    attr_accessor :params
+    attr_accessor :params, :errors
 
     # @param params [Hash] api payload
     def perform(params:)
-      self.params = params
+      self.params = params.deep_symbolize_keys
       # FIXME: add param validation and capture raw request
 
-      referral = nil
+      self.errors = []
+      success = nil
       # transact assumes we are only mutating records in the warehouse db
       HmisExternalApis::Referral.transaction do
         referral = create_referral
-        create_referral_postings(referral)
-        create_referral_household_members(referral)
+        raise ActiveRecord::Rollback unless create_referral_posting(referral)
+
+        raise ActiveRecord::Rollback unless create_referral_household_members(referral)
+
+        success = referral
       end
-      referral
+      [success, errors]
     end
 
     protected
 
     def create_referral
       params => {referral_id:, referral_date:, service_coordinator:}
-      referral = referral_scope.new
+      referral = HmisExternalApis::Referral.new
       referral.identifier = referral_id
       referral.referral_date = referral_date
       referral.service_coordinator = :service_coordinator
       referral.save!
       referral
     end
 
-    def create_referral_postings(referral)
-      posting_attrs = params.fetch(:postings)
+    def create_referral_posting(referral)
+      (posting_id, program_id, unit_type_id, referral_request_id) = params.values_at(:posting_id, :program_id, :unit_type_id, :referral_request_id)
+      raise unless posting_id && program_id # required fields, should be caught in validation
```

Comment:
We should error out on missing `unit_type_id` here

Reply:
thanks for the catch!